### PR TITLE
Ports tgstation#57081 tgui: Typescript and Jest update

### PR DESCRIPTION
**Pull request by mordent-goonstation**
Plus associated follow-up PRs.

## About the PR 

Ports:

- https://github.com/tgstation/tgstation/pull/57081 - tgui: Typescript and Jest update
- https://github.com/tgstation/tgstation/pull/57499 - Fix tgui reloading
- https://github.com/tgstation/tgstation/pull/57465 - tgui: Upgrade to Yarn 2.4.1 and TypeScript 4.2
- https://github.com/tgstation/tgstation/pull/57154 - Fix lint task in tgui powershell script
- https://github.com/tgstation/tgstation/pull/57649 - Revert "Flush queue in preloading part of tgui_panel"
- https://github.com/tgstation/tgstation/pull/57665 - Revert "Flush queue in preloading part of tgui_panel" - Part 2
- https://github.com/tgstation/tgstation/pull/57698 - Add "ESLint Fix All" on Ctrl+S to .vscode/settings.json
- https://github.com/tgstation/tgstation/pull/57746 - Typescript tgui interfaces fixups.

Partial port of:

- https://github.com/tgstation/tgstation/pull/57251 - Fix a number of build issues
- https://github.com/tgstation/tgstation/pull/57326 - Brutally fix gitignore

TODO

- [x] Resolve pnpapi error related to either space in directory or too long directory (documented error for now)

This PR is a combination of all of the above, as they all build upon each other to a stable product.

At present there is an issue when running `bin/tgui --dev` if your working directory is either too long or has a space in it (I didn't isolate which, but moving the project from `D:/Users/Firstname Lastname/Documents/Development/BYOND/goonstation` to `D:/Dev/goonstation` solved the error. The visible error is:

> Error: Unable to locate pnpapi, the module 'D:\Users\Firstname Lastname\Documents\Development\BYOND\goonstation\tgui\packages\tgui-dev-server\index.js' is controlled by multiple pnpapi instances.
> This is usually caused by using the global cache (enableGlobalCache: true)

The partial port of https://github.com/tgstation/tgstation/pull/57251 does not port any of the contents of tg's `tools` directory, as we use a different process.

The partial port of https://github.com/tgstation/tgstation/pull/57326 only touches the chunk needed to make the relevant files (contents of `tgui/.yarn/sdks/eslint/lib` and `tgui/.yarn/sdks/typescript/lib`) not be gitignored (specifically the `lib/` entry). This should be looked at in more detail to make sure we're not removing some files we shouldn't be, or if there's anything else we can also tidy up.

## Why's this needed? 

Update good.
TypeScript good.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -34,27 +34,8 @@ __pycache__/
...
-lib/
-lib64/
-parts/
-sdist/
-var/
-*.egg-info/
-.installed.cfg
-*.egg
-.cache
+# Ignore compiled linux libs in the root folder, e.g. librust_g.so
```
Comment (mordent-goonstation) [inline .gitignore:37]: @ZeWaka Could you take a look through the .gitignore changes and make sure relevant? Relevant PR for comparison: https://github.com/tgstation/tgstation/pull/57326

File: tgui/.gitignore [hunk 2]
```diff
@@ -9,3 +9,4 @@ package-lock.json
 !/.yarn/plugins
 !/.yarn/releases
 !/.yarn/sdks
+!/.yarn/lock.yml
```
Comment (mordent-goonstation) [inline tgui/.gitignore:12]: This doesn't seem to get created, but excluding from being gitgnored in case it appears.

File: .gitignore (hunk 1 above)
Comment (ZeWaka) [reply .gitignore:37]: I know for sure the cache one is needed, style was lamenting about it for *minutes*

**Review by mordent-goonstation — COMMENTED** [review]
Confirmed the error related to pnpapi occurs in tgstation if in a sibling working directory to my goonstation working directory, and not if put in a sibling working directory to my working directory at `D:/dev`, so it's not a Goonstation specific problem. @stylemistake

**Comment by ZeWaka** [comment]
revert your changes to tgui_panel.dm re: https://github.com/tgstation/tgstation/pull/57649

**Comment by stylemistake** [comment]
https://github.com/tgstation/tgstation/pull/57698

**Comment by stylemistake** [comment]
> if in a sibling working directory to my goonstation working directory, and not if put in a sibling working directory to my working directory at `D:/dev`

Uh, what?

**Comment by mordent-goonstation** [comment]
> > if in a sibling working directory to my goonstation working directory, and not if put in a sibling working directory to my working directory at `D:/dev`
> 
> Uh, what?

`D:/Users/Firstname Lastname/Documents/Development/goonstation` <- `tgui/bin --dev` breaks with error
`D:/Dev/goonstation` <- `tgui/bin --dev` works
`D:/Users/Firstname Lastname/Documents/Development/tgstation` <- `tgui/bin --dev` breaks with error
`D:/Dev/tgstation` <- `tgui/bin --dev` works

**Comment by mordent-goonstation** [comment]
> revert your changes to tgui_panel.dm re: [tgstation/tgstation#57649](https://github.com/tgstation/tgstation/pull/57649)

I *think* I'm already up-to-date here. I'll add 57649 to the PR description.

**Comment by mordent-goonstation** [comment]
General awareness with this PR: you may see a notification like the below in VSCode. People should be clicking "Allow" if they want to benefit from a better experience.

image

**Comment by mordent-goonstation** [comment]
The ESLint extension needed needs the user to complete a couple of actions, specifically ("Use NODE_PATH value defined via setting ./tgui/.yarn/sdks"):
image

And in the bottom-right ("Allow"):
image

**Comment by stylemistake** [comment]
https://github.com/tgstation/tgstation/pull/57746

File: .gitignore (hunk 1 above)
Comment (mordent-goonstation) [reply .gitignore:37]: They have `.cache` further down in their `.gitignore` under `# Unit test / coverage reports`, we don't have such a section. Running `bin/tgui --test` doesn't seem to generate any files that show in `git status`, so unsure if we need to re-add.

Comment (ZeWaka) [reply .gitignore:37]: eh whatever really

File: .github/DEV_GUIDE.md [hunk 3]
```diff
@@ -52,6 +52,7 @@ If this does not work, search for the following extensions:
  - GitHub Pull Requests - lets you make a pull request directly from the editor
  - Bracket Pair Colorizer 2 - paints matching brackets the same color
  - Git Graph - lets you work with branches visually for ease of use
+ - ESLint - lets you see some code errors in your editor as you work, also auto-fixes code style items on save
```
Comment (ZeWaka) [inline .github/DEV_GUIDE.md:55]: ok

File: .gitignore [hunk 4]
```diff
@@ -34,27 +34,8 @@ __pycache__/
...
-#*.so
-
-# Distribution / packaging
-.Python
-env/
-build/
-develop-eggs/
-dist/
-downloads/
-eggs/
```
Comment (ZeWaka) [inline .gitignore:47]: egg

File: tgui/.eslintrc.yml [hunk 5]
```diff
@@ -1,6 +1,7 @@
-parser: '@babel/eslint-parser'
+root: true
+parser: '@typescript-eslint/parser'
 parserOptions:
-  ecmaVersion: 2019
+  ecmaVersion: 2020
```
Comment (ZeWaka) [inline tgui/.eslintrc.yml:4]: 🤔

**Review by ZeWaka — APPROVED** [review]
so i got through this halfway and then decided i was hungry :shipit:

File: .github/DEV_GUIDE.md (hunk 3 above)
Comment (mordent-goonstation) [reply .github/DEV_GUIDE.md:55]: I'm actually a little torn on fix-lint-errors-on-save, but we'll see how it goes.

File: tgui/.eslintrc.yml (hunk 5 above)
Comment (mordent-goonstation) [reply tgui/.eslintrc.yml:4]: https://www.freecodecamp.org/news/javascript-new-features-es2020/

Comment (ZeWaka) [reply tgui/.eslintrc.yml:4]: it's 2021 u doofus

Comment (mordent-goonstation) [reply tgui/.eslintrc.yml:4]: I'm porting a PR, not getting us to the bleeding edge.

https://github.com/tgstation/tgstation/blob/master/tgui/.eslintrc.yml#L4

**Pull request merged**